src/lib/reasoners/sat_solver_sig.mli

```diff
@@ -31,9 +31,11 @@
 module type S = sig
   type t
 
-  exception Sat of t
-  exception Unsat of Explanation.t
-  exception I_dont_know of t
+  type res =
+    | Done of t
+    | Sat of t
+    | Unsat of Explanation.t
+    | I_dont_know of t
```

Comment:
I'd suggest using

```ocaml
type outcome = Sat of t | Unsat of Explanation.t | I_dont_know of t

type 'a res = ('a, outcome) Result.t
```

and use `t res` as return type for `assume` / `assume_th_elt` / `pred_def`, and `outcome` for `unsat`.

Reply:
Your version adds an extra pointer in the `Error` case. Besides, the `Error` constructor of `Result.t` would be badly named. It is not an error, it is a standard response.

Reply:
I don't think the extra pointer here matters, I was trying to optimise for readability —sure  `Error` is badly named, but there is a fundamental difference here between `Done` and `Sat / Unsat / I_dont_know` (as exemplified by the fact that `unsat` can only return `Sat / Unsat / I_dont_know`, never `Done`).

But actually, as mentioned in another comment, that `Sat of t | Unsat of Explanation.t | I_dont_know of t` should be enough, bcause it makes sense for `assume` / `assume_th_elt` / `pred_def` to return `I_dont_know`.

Reply:
Yes we could indeed replace `Done` by `I_dont_know` as they seem to play the same role (and I kind of did not like the `Done` name)